generate_densenet_predictions.py

```diff
@@ -114,11 +109,10 @@ def __len__(self):
 	# save predictions to h5
 	with tables.open_file(args.pytable_path, 'a') as f:
 		if 'predictions' in f.root:
-			predictions_dataset = f.root.predictions
+			print(f"Predictions key already exists in {args.pytable_path}")
 		else:
-			predictions_dataset = f.create_carray(f.root, "predictions", dtype, np.array(predictions).shape)
+			f.create_carray(f.root, "predictions", dtype, np.array(predictions).shape)
 
-		predictions_dataset[:] = predictions
```

Comment:
should this be removed?

Reply:
My mistake! Thanks for pointing it out. I fixed this bug and tested the script to make sure it works correctly. See the fixed version below.

Reply:
sorry buddy, don't think you're quite there yet

this looks like it will "only" work if "predictions" doesn't exist in the pytable.

if it *does* exist in the pytable, where is the statement which updates its values?

generally speaking, you have the wrong pattern overall here, which as you are seeing is leading you to fall into a number of unexpected bug situations

the better way of doing this is to use the inverse and check for non-existence, and make the field if needed. then  ***always*** assign the predictions value

```
if not 'predictions' in f.root:
    predictions_dataset = f.create_carray(f.root, "predictions", dtype, np.array(predictions).shape)

    
predictions_dataset[:] = predictions
print(f'Predictions have been saved to {args.pytable_path}')

```

Reply:
Okay, thanks for the feedback.

I'd like to note that the decision to make the code only "work" if "predictions" doesn't exist was intentional, for data safety. 

My main consideration was to ensure that no unsuspecting user had their data overwritten (e.g. a user generated predictions with a different model and forgot about them before running my code.)

While using the generate_densenet_predictions.py code myself, I liked that the script did nothing (and notified me) if the predictions dataset already existed. From there I could manually delete the predictions dataset, or rename the dataset ("old_densenet_predictions" maybe) before re-running the script.

Still, I can see why some users may want the script to be capable of overwriting their predictions dataset. Would it be okay if I added the ability to overwrite but only after a (Y/n) user confirmation?

Reply:
got it, i see your point

2 thoughts:

1) if someone is running a script called "generate predicions", i would wonder what their basic expectations would be? normally concerns for overwriting are as a result of potentially 'second' side effects that the user doesn't expect

for example, using the "rm" command, directly removes files? why? because that's what its supposed to do

but something like "copy" in windows, asks you, "are you sure?" when the copy will actually be an overwrite of an existing file because that is likely an unintended consequence?

so its a fair question to ask, what is the expected baseline usecase of using your script

2) "if the predictions dataset already existed. From there I could manually delete the predictions dataset, or rename the dataset ("old_densenet_predictions" maybe) before re-running the script."

This seems like a lot of extra work for someone who really just wants to generate predictions : ) 

> Would it be okay if I added the ability to overwrite but only after a (Y/n) user confirmation?

This seems like a very reasonable solution, but please note that it should go _at the top_ of the program. even in the case that you're proposing here, its customary to check first to make sure that things are "as expected" so that they can fail fast before launching into other things.

in the current approach, you're still loading the DL model (which can take a few seconds/minutes if very large) occupying the GPU, etc etc....just to quickly quit 2 lines from the bottom. 

so the right way is after parsing the arguments to see if the environment is "sane" before proceeding, this is where your check would go

make sense?

Reply:
Great. I've tested and committed a new version that notifies the user of an existing predictions dataset at the **beginning** of the script, and later overwrites the dataset.

Note that I've chosen not to perform an "overwrite" in the most literal sense, since carrays have immutable length and thus 
`f.root.predictions[:] = predictions`
might throw an indexError if for some reason the number of predictions has changed. Instead I remove the predictions dataset and create a new one with the correct shape.

And I see your first point too. My "better safe than sorry" approach here might not be necessary if the user expects the script to overwrite. I probably could have gotten away with automatically overwriting so long as I made that behavior clear in the blog post.

Reply:
So close!

You can entirely remove the "preds_exists" variable from the script. This type of flagging is usually done when the specific operation it flags is expensive to compute, e.g., "any duplicates in an unsorted listed of 1 billion numbers". in this case, you will want to do the computation a single time and save the result

in this case:

```
if 'predictions' in f.root:
```

is likely a hash lookup into a dictionary in the pytable, and will be done in O(1) time, and in practice a few milliseconds

as such, its better to _always_ directly run this condition, as opposed to putting in surrogate house keeping code to try and keep track of it, since this  can open you up to bugs later on. i.e., what if the status changes somewhere (by another script running in parallel?), and your flag variable doesn't get updated? 

the solution - > always get your data right from the horses mouth : ) 

see what i mean?

Reply:
Yes, I see what you mean. Always happy to learn better coding practice :)

I made the change and committed it.